packages/flutter_build/lib/src/kernel_builder.dart

```diff
@@ -0,0 +1,268 @@
+    final String packageName = buildStep.inputId.package;
+    final String oldPackagesContents = await File(packagesPath).readAsString();
+    // Note: currently we only replace the root package with a multiroot
+    // scheme. To support codegen on arbitrary packages we will need to do
+    // this for each dependency.
+    final String newPackagesContents = oldPackagesContents.replaceFirst('$packageName:lib/', '$packageName:$multiRootScheme:///lib/');
+    await packagesFile.writeAsString(newPackagesContents);
+
+    // start up the frontend server with configuration.
+    final List<String> arguments = <String>[
+      frontendServerPath,
+      '--sdk-root',
+      sdkRoot,
+      '--strong',
+      '--target=flutter',
+    ];
+    if (trackWidgetCreation) {
+      arguments.add('--track-widget-creation');
+    }
+    if (!linkPlatformKernelIn) {
+      arguments.add('--no-link-platform');
+    }
+    if (aot) {
+      arguments.add('--aot');
+      arguments.add('--tfa');
+    }
+    if (targetProductVm) {
+      arguments.add('-Ddart.vm.product=true');
+    }
+    if (incrementalCompilerByteStorePath != null) {
+      arguments.add('--incremental');
+    }
+    final String generatedRoot = path.join(projectDir.absolute.path, '.dart_tool', 'build', 'generated', '$packageName');
+    arguments.addAll(<String>[
+      '--packages',
+      packagesFile.path,
+      '--output-dill',
+      outputFile.path,
+      '--filesystem-root',
+      projectDir.absolute.path,
+      '--filesystem-root',
+      generatedRoot,
+      '--filesystem-scheme',
+      multiRootScheme,
+    ]);
+    if (extraFrontEndOptions != null) {
+      arguments.addAll(extraFrontEndOptions);
+    }
+    final Uri mainUri = PackageUriMapper.findUri(
+      mainPath,
+      packagesFile.path,
+      multiRootScheme,
+      <String>[projectDir.absolute.path, generatedRoot],
+    );
+    arguments.add(mainUri.toString());
+    // Invoke the frontend server and copy the generated dill back to the
+    // generated directory.
+    try {
+      final Process server = await Process.start(engineDartBinaryPath, arguments);
+      final StdoutHandler _stdoutHandler = StdoutHandler();
+      server.stderr.transform<String>(utf8.decoder).listen(log.shout);
+      server.stdout
+        .transform<String>(utf8.decoder)
+        .transform<String>(const LineSplitter())
+        .listen(_stdoutHandler.handler);
+      await server.exitCode;
+      await _stdoutHandler.compilerOutput.future;
+      await buildStep.writeAsBytes(outputId, await outputFile.readAsBytes());
+      await buildStep.writeAsBytes(packagesOutputId, await packagesFile.readAsBytes());
+    } catch (err, stackTrace) {
+      log.shout('frontend server failed to start: $err, $stackTrace');
+    }
+  }
+}
+
+class StdoutHandler {
+  StdoutHandler() {
+    reset();
+  }
+
+  bool compilerMessageReceived = false;
+  String boundaryKey;
+  Completer<CompilerOutput> compilerOutput;
```

Comment:
I think these fields can be private. They are not used outside this class. Or maybe the whole class should be private?